src/formatters/mdapi/retrieveResultFormatter.ts

```diff
@@ -0,0 +1,138 @@
+    this.options.zipFileName ??= 'unpackaged.zip';
+    const zipFilePath = join(options.retrieveTargetDir, options.zipFileName);
+    this.result = Object.assign({}, result.response, { zipFilePath });
+
+    // grab warnings
+    const warnMessages = get(result, 'response.messages', []) as RetrieveMessage | RetrieveMessage[];
+    this.warnings = toArray(warnMessages);
+  }
+
+  /**
+   * Get the JSON output from the RetrieveResult.
+   *
+   * @returns RetrieveCommandResult
+   */
+  public getJson(): RetrieveCommandResult {
+    return this.result;
+  }
+
+  /**
+   * Displays retrieve results in human format.
+   */
+  public display(): void {
+    if (this.hasStatus(RequestStatus.InProgress)) {
+      // The command timed out
+      const commandWaitTime = getNumber(this.options, 'waitTime', 1440);
+      this.ux.log(messages.getMessage('retrieveTimeout', [commandWaitTime]));
+      return;
+    }
+
+    if (this.isSuccess()) {
+      this.ux.log(`Wrote retrieve zip to ${this.result.zipFilePath}`);
+      if (this.options.unzip) {
+        const extractPath = join(this.options.retrieveTargetDir, parse(this.options.zipFileName).name);
+        this.ux.log(`Extracted ${this.options.zipFileName} to: ${extractPath}`);
+      }
+      if (this.options.verbose) {
+        const retrievedFiles = toArray(this.result.fileProperties);
+        if (retrievedFiles?.length) {
+          this.displaySuccesses(retrievedFiles);
+        }
+      }
+      if (this.warnings.length) {
+        this.displayWarnings();
+      }
+    } else {
+      this.displayErrors();
+    }
+  }
+
+  protected hasStatus(status: RequestStatus): boolean {
+    return this.result?.status === status;
+  }
+
+  private displayWarnings(): void {
+    this.ux.styledHeader(yellow(messages.getMessage('retrievedSourceWarningsHeader')));
+    const columns = [
+      { key: 'fileName', label: 'FILE NAME' },
+      { key: 'problem', label: 'PROBLEM' },
+    ];
+    this.ux.table(this.warnings, { columns });
+    this.ux.log();
+  }
+
+  private displaySuccesses(retrievedFiles: FileProperties[]): void {
+    this.sortFileProperties(retrievedFiles);
+
+    this.ux.log('');
+    this.ux.styledHeader(blue(`Components Retrieved [${retrievedFiles.length}]`));
+    const columns = [
+      { key: 'type', label: 'TYPE' },
+      { key: 'fileName', label: 'FILE' },
+      { key: 'fullName', label: 'NAME' },
+      { key: 'id', label: 'ID' },
+    ];
+    this.ux.table(retrievedFiles, { columns });
+  }
+
+  private displayErrors(): void {
+    // an invalid packagename retrieval will end up with a message in the `errorMessage` entry
+    const errorMessage = get(this.result, 'errorMessage') as string;
+    if (errorMessage) {
+      throw new SfdxError(errorMessage);
+    }
+    const unknownMsg: RetrieveMessage[] = [{ fileName: 'unknown', problem: 'unknown' }];
+    const responseMsgs = get(this.result, 'messages', unknownMsg) as RetrieveMessage | RetrieveMessage[];
```

Comment:
same question here about not trusting SDR typing?

If it really isn't reliable, it's worth commenting on why and creating a WI to fix.